forc/src/pkg.rs

```diff
@@ -0,0 +1,706 @@
+    compilation_order: &[NodeIx],
+) -> Result<PathMap, String> {
+    let mut path_map = PathMap::new();
+
+    // We resolve all paths in reverse compilation order.
+    // That is, we follow paths starting from the project root.
+    let mut path_resolve_order = compilation_order.iter().cloned().rev();
+
+    // Add the project's package to the map.
+    let proj_node = path_resolve_order
+        .next()
+        .ok_or_else(|| "graph must contain at least the project node".to_string())?;
+    let proj_id = graph[proj_node].id();
+    path_map.insert(proj_id, proj_manifest_dir.to_path_buf());
+
+    // Resolve all following dependencies, knowing their parents' paths will already be resolved.
+    for dep_node in path_resolve_order {
+        let dep = &graph[dep_node];
+        let dep_path = match &dep.source {
+            SourcePinned::Git(git) => {
+                git_commit_path(&dep.name, &git.source.repo, &git.commit_hash)
+            }
+            SourcePinned::Path => {
+                let parent_node = graph
+                    .edges_directed(dep_node, Direction::Incoming)
+                    .next()
+                    .ok_or_else(|| "more than one root package detected in graph".to_string())?
+                    .source();
+                let parent = &graph[parent_node];
+                let parent_path = &path_map[&parent.id()];
+                let parent_manifest = read_manifest(parent_path)?;
+                let detailed = parent_manifest
+                    .dependencies
+                    .as_ref()
+                    .and_then(|deps| match &deps[&dep.name] {
+                        Dependency::Detailed(detailed) => Some(detailed),
+                        Dependency::Simple(_) => None,
+                    })
+                    .ok_or_else(|| format!("missing path info for dependency: {}", dep.name))?;
+                let rel_dep_path = detailed
+                    .path
+                    .as_ref()
+                    .ok_or_else(|| format!("missing path info for dependency: {}", dep.name))?;
+                parent_path.join(rel_dep_path)
+            }
+            SourcePinned::Registry(_reg) => {
+                return Err("registry dependencies are not yet supported".to_string());
+            }
+        };
+        if !dep_path.exists() {
+            return Err("pinned dependency is missing local source: try `forc update`".to_string());
+        }
+        path_map.insert(dep.id(), dep_path);
+    }
+
+    Ok(path_map)
+}
+
+/// Fetch all depedencies and produce the dependency graph along with a map from each node's unique
+/// ID to its local fetched path.
+///
+/// This will determine pinned versions and commits for remote dependencies during traversal.
+pub(crate) fn fetch_deps(
+    proj_manifest_dir: PathBuf,
+    proj_manifest: &Manifest,
+    offline_mode: bool,
+) -> Result<(Graph, PathMap), String> {
+    let mut graph = Graph::new();
+    let mut path_map = PathMap::new();
```

Comment:
Consider not building the `PathMap` while constructing the `Graph`, and instead construct it afterwards using the new `graph_to_path_map` function above.

Reply:
It makes sense to collect `PathMap` as we go as it doubles as a cache for already visited dependencies.